types/yaireo__tagify/index.d.ts

```diff
@@ -1304,7 +1324,7 @@ declare class Tagify<T extends Tagify.BaseTagData = Tagify.TagData> {
     /**
      * The current settings of this tagify instance.
      */
-    settings: Tagify.TagifySettings<T>;
+    settings: Required<Tagify.TagifySettings<T>>;
```

Comment:
Can we be certain all properties actually exist in a constructed tagify instance? A few properties such as [originalInputValueFormat](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/71b4fe03978e8b722630b71a7d8a14728d3d9752/types/yaireo__tagify/index.d.ts#L817) or [templates](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/71b4fe03978e8b722630b71a7d8a14728d3d9752/types/yaireo__tagify/index.d.ts#L756) don't seem to be included in the [defaults](https://github.com/yairEO/tagify/blob/9bc961b79105bf471b3cc9bf9f0f657f968aca6c/src/parts/defaults.js) (but could be initializedsomewhere else?)

Other than that, this looks good to me.

Reply:
The templates are also set as part of the settings: https://github.com/yairEO/tagify/blob/master/src/tagify.js#L99-L102

But you are right with `originalInputValueFormat`. This can be possibly undefined. I could change it to something like
```
settings: Omit<Required<Tagify.TagifySettings<T>>, "originalInputValueFormat"> & Pick<Tagify.TagifySettings<T>, "originalInputValueFormat">
```
I need to go through each setting if they are part of the defaults or not. Maybe it is better to define seperate interfaces for settings with defaults and settings without defaults, which are then unified differently for constructor parameter and instance property?

Reply:
Personally, I feel like this may be unintentional, perhaps the author just forgot to add a default for that setting? (since pretty much every other settings seems to have a default)

> Maybe it is better to define separate interfaces for settings with defaults and settings without defaults, which are then unified differently for constructor parameter and instance property?

Good idea, Separate interfaces sound more maintainable and readable to me. And they can be reused, if that becomes necessary.

Something like this?

```ts
interface TagifyRequiredSettings {
  templates?: ...;
}
interface TagifyOptionalSettings {
  originalInputValueFormat?: ...;
}
interface TagifySettings extends TagifyRequiredSettings, TagifyOptionalSettings {
  
}
interface TagifyInstanceSettings extends Required<TagifyRequiredSettings>, TagifyOptionalSettings {  
}
interface Tagify {
  settings: TagifyInstanceSettings
}
```

Reply:
Another option could be that TagifySettings are non-optional with respect to having a default or not. So `duplicates?: boolean | undefined;` becomes `duplicates: boolean;`, but `originalInputValueFormat?: ((value: T[]) => string) | undefined;` stays the same. Then, the constructor settings are just partial `Partial<...>`. This way, we don't need to split up the settings that much. However, that would also effect the sub-settings like DropDownSettings. It wouldn't be very backward compatible, I guess.

Reply:
Indeed, `Partial<...>` and `Required<...>` aren't recursive. What I mean with the example of `DropDownSettings` is that people also might already using it and changing them to not having `undefined` as well might break it. So, this "Partial" approach would be needed for all sub-settings as well. So, we would also need to define something like:

```ts
interface DefaultDropDownSettings {
  searchKeys: string[];
  ...
}

type DropDownSettings = Partial<DefaultDropDownSettings>; // or as interface

interface DefaultTagifySettings { 
  duplicates: boolean;
  originalInputValueFormat?: ((value: T[]) => string) | undefined;
  dropdown: DefaultDropDownSettings;
}

interface TagifySettings extends RecursivePartial<DefaultTagifySettings> {}
```
https://stackoverflow.com/questions/41980195/recursive-partialt-in-typescript

Or we will just leave it as it is and you need to access settings like `tagify.settings.dropdown?.maxItems?` ?

Reply:
> Or we will just leave it as it is and you need to access settings like tagify.settings.dropdown?.maxItems? ?

Yeah, let's keep it simple and leave the drop down settings as they are. If we changed this, we'd also have to check each drop down settings and make sure it really cannot be undefined. Better to err on the safe side. And I don't think anybody would be accessing the drop down settings so often the additional optional chaining checks could affect the performance.